### ISSUE TITLE
Making some uses of `@__DIR__` relocatable

### ISSUE THREAD

**Issue by gbaraldi** [JuliaLang/julia#54430]
The proposed used I have for this is for when packages are looking for their Artifacts.toml file, they have to annoying things, and non relocatable ones even more annyoingly. (Maybe we need to add something like `@__PKGDIR` that does the relocatable thing.

**Comment by topolarity** [comment]
I agree with @gbaraldi that this is a serious limitation. It means that most uses of `@__DIR__` cause your package to become immediately non-relocatable.

There are two common exceptions that are relocatable:
 - the `@__DIR__` path is used to access files entirely at parse-time / compile-time - e.g. JLL's do this: https://github.com/JuliaLang/julia/blob/4793328656cc1bffaa5625a62b7cc36baa32760c/stdlib/Artifacts/src/Artifacts.jl#L691-L693
 - the `@__DIR__` path is never used to open/write any files at all (just printing/logging) - `@error` does this

https://github.com/JuliaLang/julia/pull/53906 added a `Base.isrelocatable` predicate, but it is not able to detect this problem.

**Comment by topolarity** [comment]
> The issue arises if one tries to move that package to another place, either via PkgCompiler or via copying the depot around.

There are more challenges for PackageCompiler.jl too.

It deletes any source files from its relocated "depot". So even if this is relocatable, it would need to have a way to detect a dependency on package source so that it can copy those files to make this work.

**Comment by KristofferC** [comment]
https://github.com/JuliaPackaging/RelocatableFolders.jl is a bit of a "hack" to work around this.

**Comment by fatteneder** [comment]
FWIW Pkg.jl [best practices](https://pkgdocs.julialang.org/v1/creating-packages/#Best-Practices) already mention that one should not assume a package to be located in a stable location. Using the current `@__DIR__` obviously violates that.

**Comment by KristofferC** [comment]
What should be done for this issue to be solved? It seems to lack a clear goal right now.

**Comment by fatteneder** [comment]
I personally would not bother with making uses of `@__DIR__` relocatable. I have used things like `normpath(joinpath(@__DIR__, <some relative path>))` myself before, and I tinkering with `@__DIR__` would only break those uses.

Instead I would add another macro like, something like `@__PKGDIR__` as suggested by Gabriel initially.
